Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift

```diff
@@ -318,8 +315,28 @@ extension HTTPConnectionPool {
         private mutating func nextActionForFailedConnection(at index: Int, on eventLoop: EventLoop) -> Action {
             switch self.state {
             case .running:
-                let hasPendingRequest = !self.requests.isEmpty(for: eventLoop) || !self.requests.isEmpty(for: nil)
-                guard hasPendingRequest else {
+                // we do not know if we have created this connection for a request with a required
+                // event loop or not. However, we do not need this information and can infer
+                // if we need to create a new connection because we will only ever create one connection
+                // per event loop for required event loop requests and only need one connection for
+                // general purpose requests.
+
+                // we need to start a new on connection in two cases:
+                let needGeneralPurposeConnection =
+                    // 1. if we have general purpose requests
+                    !self.requests.isEmpty(for: nil) &&
+                    // and no connection starting or active
+                    !self.connections.hasStartingOrActiveConnection()
+
+                let needRequiredEventLoopConnection =
+                    // 2. or if we have requests for a required event loop
+                    !self.requests.isEmpty(for: eventLoop) &&
```

Comment:
This is true for `self.connections` but we do not really search `self.requests` twice because the requests with required event loop and without required event loop live in separate properties inside of `self.requests`.